### ISSUE TITLE
Cgen error with sum types

### ISSUE THREAD

**Issue by jaddison06** [vlang/v#9311]
**V version:** V 0.2.2 818be80
**OS:** linux, Pop!_OS 20.10

**What did you do?**
Refactored some of my functions. I had a RichText struct which contained a load of data about formatting, but it was getting tiresome to write out the struct syntax every time I wanted some plain text, so I created a new sum type.
```v
type Text = RichText | string
```
I changed all my functions etc to work with this instead of the old RichText type

**What did you expect to see?**
The code operating as normal
 
**What did you see instead?**
```
/tmp/v/refactor.6112290544865777190.tmp.c:2702: error: invalid operand types for binary operation
/tmp/v/refactor.6112290544865777190.tmp.c:2702: error: invalid aggregate type for register load
...
(Use `v -cg` to print the entire error message)

builder error: 
C error. This should never happen.
```

I don't know what's going on - I can link my code if needed.

**Comment by jaddison06** [comment]
Did a bit more digging. Here's the offending C code:
```c

static bool main__MenuItem_struct_eq(main__MenuItem a, main__MenuItem b) {
	if (a.title != b.title) { // line 2767
		return false;
	}
	if (a.highlight_on_hover != b.highlight_on_hover) {
		return false;
	}
	if (!term__ui__Color_struct_eq(a.hover_col, b.hover_col)) {
		return false;
	}
	return true;
}
```
MenuItem.title is of type Text, so it looks like something's wrong with the comparison between unknown types.

**Comment by jaddison06** [comment]
Tried doing it manually like this:
```v
fn (a Text) == (b Text) bool {
	if a is string || b is string {
		return a.str() == b.str()
	} else if a is RichText && b is RichText {
		return (a as RichText) == (b as RichText)
	}
}
```
Compiler doesn't like that either - fair enough:
```
./menu.v:74:1: error: operator methods are only allowed for struct and type alias
   72 | }
   73 | 
   74 | fn (a Text) == (b Text) bool {
      | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   75 |     if a is string || b is string {
   76 |         return a.str() == b.str()
```

**Comment by Delta456** [comment]
```v
if (a.title != b.title) {
		return false;
}
```

This should not be generated as you can't compare sum types like that. cc @yuyi98

**Comment by spaceface777** [comment]
we need to generate `_eq` functions for sumtypes as well in cgen:
```v
fn (a MySumType) eq (b MySumType) bool {
    if a._typ != b._typ { return false }
    switch a._typ {
        1 { return Subtype1__eq(a._Subtype1, b._Subtype1) }
        2 { return Subtype2__eq(a._Subtype2, b._Subtype2) }
        (...)
    }
}
```

**Comment by jaddison06** [comment]
@spaceface777 Or allow people to overload ==. At the moment the compiler doesn't allow it, but I can think of times where I want to compare a RichText to a string. Your implementation would be a good default, but I think customisability is important.

**Comment by Dialga** [comment]
Fixed by #9408 0258482caf9a73b733d428fa683c054f6cc27510.
